notebooks/02_model/hypertune_spark_deep_dive.ipynb

```diff
@@ -0,0 +1,983 @@
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "model.explainParams()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "[1.3818693574087881,\n",
+       " 1.03881169247089,\n",
+       " 1.4149480500849774,\n",
+       " 1.8957753237837804,\n",
+       " 2.707401263212467,\n",
+       " 1.3818693597518183,\n",
+       " 1.022026009040587,\n",
+       " 1.4517896758192947,\n",
+       " 1.9079444576340328,\n",
+       " 2.786421555547063]"
+      ]
+     },
+     "execution_count": 12,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "model.validationMetrics"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To get the best model, just do"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model_best_spark = model.bestModel"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 3 Hyperparameter tuning with `hyperopt` in sequential mode"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "`hyperopt` is an open source Python package that is designed for tuning parameters for generic function with any pre-defined loss. More information about `hyperopt` can be found [here](https://github.com/hyperopt/hyperopt).\n",
+    "\n",
+    "In `hyperopt`, an *objective* function is defined for optimizing the hyper parameters. In this case, the objective is similar to that in the Spark native construct situation, which is *to the RMSE metric for an ALS recommender*. Parameters of `rank` and `regParam` are used as hyperparameters. \n",
+    "\n",
+    "The objective function shown below demonstrates a RMSE loss for an ALS recommender. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Customize an objective function\n",
+    "def objective(params):\n",
+    "    rank = params['rank']\n",
+    "    reg = params['reg']\n",
+    "    train = params['train'] \n",
+    "    valid = params['valid'] \n",
+    "    col_user = params['col_user'] \n",
+    "    col_item = params['col_item']\n",
+    "    col_rating = params['col_rating'] \n",
+    "    col_prediction = params['col_prediction'] \n",
+    "    k = params['k']\n",
+    "    relevancy_method = params['relevancy_method']\n",
+    "    \n",
+    "    als = ALS(\n",
+    "        rank=rank,\n",
+    "        maxIter=15,\n",
+    "        implicitPrefs=False,\n",
+    "        alpha=0.1,\n",
+    "        regParam=reg,\n",
+    "        coldStartStrategy='drop',\n",
+    "        nonnegative=False,\n",
+    "        seed=0,\n",
+    "        **HEADER_ALS\n",
+    "    )\n",
+    "    \n",
+    "    model = als.fit(train)\n",
```

Comment:
Here we are using train (60% of the original data) and valid (20%). In the previous method, `TrainValidationSplit` is instantiated to use train, which is splitted internally in 80%-20%. Is this correct? If so, in the first method you should input train+valid.

Reply:
Good catch. I modified the split ratio.